<a href="https://colab.research.google.com/github/Kwasi-Dankwa/Easy_Visa_CS/blob/main/AI%26DS_Week8_Project2_Full_Code_Notebook_EasyVisa_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Problem Statement**

## Context

Business communities in the United States are facing high demand for human resources, but one of the constant challenges is identifying and attracting the right talent, which is perhaps the most important element in remaining competitive. Companies in the United States look for hard-working, talented, and qualified individuals both locally and abroad.

The Immigration and Nationality Act (INA) of the US permits foreign workers to come to the United States to work on either a temporary or permanent basis. The act also protects US workers against adverse impacts on their wages or working conditions by ensuring US employers' compliance with statutory requirements when they hire foreign workers to fill workforce shortages. The immigration programs are administered by the Office of Foreign Labor Certification (OFLC).

OFLC processes job certification applications for employers seeking to bring foreign workers into the United States and grants certifications in cases where employers can demonstrate that there are not sufficient US workers available to perform the work at wages that meet or exceed the wage paid for the occupation in the area of intended employment.

## Objective

In FY 2016, the OFLC processed 775,979 employer applications for 1,699,957 positions for temporary and permanent labor certifications, a nine percent increase over the previous year. Reviewing every case is becoming a tedious task as the number of applicants increases every year.

The growing number of applicants calls for a machine learning based solution that can help shortlist candidates with higher chances of visa approval. OFLC has hired the firm EasyVisa for data-driven solutions. As a data scientist at EasyVisa, you must analyze the data provided and, with the help of a classification model:

1. Facilitate the process of visa approvals.
2. Recommend a suitable profile for applicants whose visa should be certified or denied, based on the drivers that significantly influence the case status.

## Data Description

The data contains attributes of the employee and the employer. The detailed data dictionary is given below.

* **case_id**: ID of each visa application
* **continent**: Information on the continent of the employee
* **education_of_employee**: Information on the education of the employee
* **has_job_experience**: Does the employee have any job experience? Y = Yes; N = No
* **requires_job_training**: Does the employee require any job training? Y = Yes; N = No
* **no_of_employees**: Number of employees in the employer's company
* **yr_of_estab**: Year in which the employer's company was established
* **region_of_employment**: Information on the foreign worker's intended region of employment in the US
* **prevailing_wage**: Average wage paid to similarly employed workers in a specific occupation in the area of intended employment. The purpose of the prevailing wage is to ensure that the foreign worker is not underpaid compared to other workers offering the same or similar service in the same area of employment
* **unit_of_wage**: Unit of prevailing wage. Values include Hourly, Weekly, Monthly, and Yearly
* **full_time_position**: Is the position of work full-time? Y = Full-Time Position; N = Part-Time Position
* **case_status**: Flag indicating whether the visa was certified or denied

## **Note: This is a sample solution for the project. Projects will NOT be graded on the basis of how well the submission matches this sample solution. Projects will be graded on the basis of the rubric only.**

# **Importing necessary libraries**

In [1]:
# Installing the libraries with the specified version.
%pip install -q pandas==2.2.2 numpy==2.0.2 matplotlib==3.10.0 seaborn==0.13.2 scikit-learn==1.6.1

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.4/4.4 MB 89.9 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  Preparing metadata (pyproject.toml) ... error
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.


**Note**:
* After running the above cell, kindly restart the runtime (for Google Colab) or notebook kernel (for Jupyter Notebook), and run all cells sequentially from the next cell.
* On executing the above line of code, you might see a warning regarding package dependencies. This error message can be ignored as the above code ensures that all necessary libraries and their dependencies are maintained to successfully execute the code in **this notebook**.

In [6]:
# Import Libraries
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split

# **Loading the dataset**

In [5]:
import pandas as pd
df = pd.read_csv('EasyVisa.csv')

In [8]:
# Viewing dataset
df.head()

,case_id,continent,education_of_employee,has_job_experience,requires_job_training,no_of_employees,yr_of_estab,region_of_employment,prevailing_wage,unit_of_wage,full_time_position,case_status
0,EZYV01,Asia,High School,N,N,14513,2007,West,592.2029,Hour,Y,Denied
1,EZYV02,Asia,Master's,Y,N,2412,2002,Northeast,83425.6500,Year,Y,Certified
2,EZYV03,Asia,Bachelor's,N,Y,44444,2008,West,122996.8600,Year,Y,Denied
3,EZYV04,Asia,Bachelor's,N,N,98,1897,West,83434.0300,Year,Y,Denied
4,EZYV05,Africa,Master's,Y,N,1082,2005,South,149907.3900,Year,Y,Certified


In [11]:
df.shape

(25480, 12)

Checking the data types of the columns

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25480 entries, 0 to 25479
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   case_id                25480 non-null  object 
 1   continent              25480 non-null  object 
 2   education_of_employee  25480 non-null  object 
 3   has_job_experience     25480 non-null  object 
 4   requires_job_training  25480 non-null  object 
 5   no_of_employees        25480 non-null  int64  
 6   yr_of_estab            25480 non-null  int64  
 7   region_of_employment   25480 non-null  object 
 8   prevailing_wage        25480 non-null  float64
 9   unit_of_wage           25480 non-null  object 
 10  full_time_position     25480 non-null  object 
 11  case_status            25480 non-null  object 
dtypes: float64(1), int64(2), object(9)
memory usage: 2.3+ MB


# **Overview of the Dataset**

### **Observations & Sanity Checks**

* **Shape**: The dataset contains **25,480 rows and 12 columns**.
* **Data Types**: There are **9 object columns** (categorical variables), **2 integer columns** (`no_of_employees`, `yr_of_estab`), and **1 float column** (`prevailing_wage`).
* **Missing Values**: According to `df.info()`, every column has **25,480 non-null values**, indicating there are **no missing values** in the dataset.
* **Identifier**: The `case_id` column acts as a unique identifier for each application.
* **Target Variable**: `case_status` is the target categorical variable indicating if a visa application was **Certified** or **Denied**.

# **Exploratory Data Analysis (EDA)**

- EDA is an important part of any project involving data.
- It is important to investigate and understand the data better before building a model with it.
- A few questions have been mentioned below which will help you approach the analysis in the right manner and generate insights from the data.
- A thorough analysis of the data, in addition to the questions mentioned below, should be done.

**Leading Questions**


1. What is the distribution of visa case statuses (certified vs. denied)?
2. How does the education level of employees impact visa approval rates?
3. Is there a significant difference in visa approval rates between employees with and without prior job experience?
4. How does the prevailing wage affect visa approval? Do higher wages lead to higher chances of approval?
5. Do certain regions in the US have higher visa approval rates compared to others?
6. How does the number of employees in a company influence visa approval? Do larger companies have a higher approval rate?
7. Are visa approval rates different across various continents of employees? Which continent has the highest and lowest approval rates?


**[IMPORTANT]** Beyond the Basics: Please note that these are guiding questions only. To receive full points for this rubric section, you are expected to perform a thorough analysis that goes beyond these specific questions to uncover deeper trends and relationships within the data.

# **Data Pre-processing**

- Missing value treatment (check if needed)
- Outlier detection and treatment (check if needed)
- Feature engineering (check if needed)
- Preparing the data for modeling, including encoding, the train-test split, and scaling where needed
- Any other preprocessing steps (check if needed)

# **Model Building: Logistic Regression, Decision Tree, and Random Forest**

- Define the model evaluation criterion for this business problem by identifying the cost of a false positive and a false negative, and choose an appropriate primary metric.
- Build a Logistic Regression model on the scaled data.
- Build a Decision Tree model on the encoded data.
- Build a Random Forest model on the encoded data.
- Evaluate each model on the training data and the test data, and comment on overfitting or the generalization gap.

# **Model Performance Comparison and Final Model Selection**

- Compare the training and test performance of all three models on your chosen primary metric, along with Accuracy, Precision, and Recall.
- Identify which model overfits the most and which model generalizes best.
- Select a final model with a clear, metric-based justification tied to the business problem.
- Report the final model's performance on the test data.

# **Actionable Insights and Recommendations**



*   




___